Ethereum Name Service

SionoiS opened the pull request:
Please advise on error handling and test cases.

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+type ContractError = crate::contract::Error;
+type EthError = crate::ethabi::Error;
+
+use hex::ToHex;
+
+use idna::Config;
```
Could you move the `use` statements all together? Before the type definitions.

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+}
+
+impl<T: Transport> Ens<T> {
+    //https://docs.ens.domains/contract-api-reference/name-processing#normalising-names
```
```suggestion
    // https://docs.ens.domains/contract-api-reference/name-processing#normalising-names
```

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+const TEXT_INTERFACE_ID: &[u8; 4] = &[0x59, 0xd1, 0xd4, 0x3c];
+const CONTENTHASH_INTERFACE_ID: &[u8; 4] = &[0xbc, 0x1c, 0x58, 0xd1];
+
+/// `Eth` namespace, ens
```
```suggestion
/// ENS contract interface.
```

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+//! `Eth` namespace, ens
```
I think the file should be moved to `contract` sub-module, since it's a higher-level API to the ENS contract as far as I understand.

SionoiS replied:
Let me know what you think of the new organization.

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+    }
+
+    /// Returns the owner of a name.
+    pub async fn get_owner(&self, domain: &str) -> Result<Address, ContractError> {
```
```suggestion
    pub async fn owner(&self, domain: &str) -> Result<Address, ContractError> {
```
I'd drop the verbs from method names when it's not necessary, since that's not a Rust convention.

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+    }
+
+    /// Returns the caching TTL (time-to-live) of a name.
+    pub async fn get_ttl(&self, domain: &str) -> Result<u64, ContractError> {
```
```suggestion
    pub async fn ttl(&self, domain: &str) -> Result<u64, ContractError> {
```

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+
+impl<T: Transport> Ens<T> {
+    //https://docs.ens.domains/contract-api-reference/name-processing#normalising-names
+    fn normalize_name(&self, domain: &str) -> Result<String, ContractError> {
```
```suggestion
    fn normalize_name(&self, domain: &str) -> contract::Result<String> {
```

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+        let resolver_addr = self.registry.get_resolver(node).await?;
+        let resolver = ReverseResolver::new(self.web3.eth(), resolver_addr);
+
+        /* if !resolver.check_interface_support(*NAME_INTERFACE_ID).await? {
+            return Err(ContractError::Abi(EthError::InvalidData));
+        } */
+
```
Please remove if not needed.

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+}
+
+#[derive(Debug, Clone)]
+pub struct Resolver<T: Transport> {
```
Missing docs

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+        self.contract.query("text", (node, key), None, options, None).await
+    }
+
+    //interfaceImplementer
```
```suggestion
    // interfaceImplementer
```

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+            .await
+    }
+
+    //multicall
```
is it missing?

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+}
+
+#[derive(Debug, Clone)]
+pub struct ReverseResolver<T: Transport> {
```
docs missing

tomusdrw commented on src/api/eth_ens.rs:
```diff
@@ -0,0 +1,827 @@
+    /// Sets the resolver contract address of a name.
+    pub async fn set_resolver(
+        &self,
+        from: Address,
+        domain: &str,
+        address: Address,
```
Could we pass references instead?
```suggestion
        from: &Address,
        domain: &str,
        address: &Address,
```

SionoiS replied:
Cloning would be needed. Either the user clone() before passing or we do it for each method

tomusdrw commented on src/contract/res/ENSRegistry.json:
```diff
@@ -0,0 +1 @@
+[{"inputs":[{"internalType":"contract ENS","name":"_old","type":"address"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"owner","type":"address"},{"indexed":true,"internalType":"address","name":"operator","type":"address"},{"indexed":false,"internalType":"bool","name":"approved","type":"bool"}]," ...
```
Can we include a version of the API somewhere or some instructions where to fetch the new ABI?

SionoiS replied:
For the minified version I used the fonctionalities of etherscan.io but ENS [github](https://github.com/ensdomains/ens-contracts) have the contract 

Will add a comment.

tomusdrw reviewed (commented):
Thank you so much for the contribution.

I don't know details of ENS, but I trust it's implemented correctly. Regarding the tests it should be possible to verify payload sent to `eth_call`.

It would be good to polish docs as well.

I think the error handling is okay, unless there are cases where specific reversion reasons from EVM are useful for the end users.

SionoiS commented:
I change the organisation and added missing functions even those without info and added links to contracts.

Tests, example and docs will have to wait a bit.

tomusdrw commented on src/contract/ens/public_resolver.rs:
```diff
@@ -0,0 +1,258 @@
+
+impl<T: Transport> PublicResolver<T> {
+    // https://docs.ens.domains/contract-api-reference/publicresolver#get-contract-abi
+    pub async fn _abi(&self, node: NameHash, content_types: U256) -> Result<(U256, Vec<u8>), ContractError> {
```
Can you remove the leading `_` from all the names? :)

tomusdrw commented on src/contract/ens/public_resolver.rs:
```diff
@@ -0,0 +1,258 @@
+}
+
+impl<T: Transport> PublicResolver<T> {
+    // https://docs.ens.domains/contract-api-reference/publicresolver#get-contract-abi
```
```suggestion
    /// https://docs.ens.domains/contract-api-reference/publicresolver#get-contract-abi
```

Please use rust doc comment format everywhere. Ideally we should have a one sentence description, followed by an empty line before the link as well.

tomusdrw commented on src/contract/ens/registry.rs:
```diff
@@ -0,0 +1,183 @@
+
+const ENS_REGISTRY_ADDRESS: &str = "00000000000C2E074eC69A0dFb2997BA6C7d2e1e";
+
+// See https://github.com/ensdomains/ens/blob/master/contracts/ENS.sol for the contract interface.
```
```suggestion
/// See https://github.com/ensdomains/ens/blob/master/contracts/ENS.sol for the contract interface.
```

tomusdrw replied:
Same in this file.

tomusdrw commented on src/contract/ens/reverse_resolver.rs:
```diff
@@ -0,0 +1,47 @@
+
+type ContractError = crate::contract::Error;
+
+// See https://github.com/ensdomains/resolvers/blob/master/contracts/DefaultReverseResolver.sol for contract interface.
```
```suggestion
/// See https://github.com/ensdomains/resolvers/blob/master/contracts/DefaultReverseResolver.sol for contract interface.
```
and here

tomusdrw commented on src/contract/ens/eth_ens.rs:
```diff
@@ -0,0 +1,409 @@
+
+impl<T: Transport> Ens<T> {
+    /// Normalize a domain name for namehash processing.
+    // https://docs.ens.domains/contract-api-reference/name-processing#normalising-names
```
```suggestion
    ///
    /// https://docs.ens.domains/contract-api-reference/name-processing#normalising-names
```

tomusdrw commented on src/contract/ens/eth_ens.rs:
```diff
@@ -0,0 +1,409 @@
+            .await
+    }
+
+    /// Sets or clears an approval.
```
```suggestion
    /// Sets or clears an approval.
    /// 
```

tomusdrw commented on src/contract/ens/eth_ens.rs:
```diff
@@ -0,0 +1,409 @@
+        self.registry.check_approval(owner, operator).await
+    }
+
+    /// Returns true if node exists in this ENS registry.
+    /// This will return false for records that are in the legacy ENS registry but have not yet been migrated to the new one.
```
```suggestion
    /// Returns true if node exists in this ENS registry.
    ///
    /// This will return false for records that are in the legacy ENS registry but have not yet been migrated to the new one.
```

tomusdrw commented on src/contract/ens/public_resolver.rs:
```diff
@@ -0,0 +1,258 @@
+
+type ContractError = crate::contract::Error;
+
+// See https://github.com/ensdomains/resolvers/blob/master/contracts/Resolver.sol for resolver interface.
```
```suggestion
/// See https://github.com/ensdomains/resolvers/blob/master/contracts/Resolver.sol for resolver interface.
```

tomusdrw commented on src/contract/ens/public_resolver.rs:
```diff
@@ -0,0 +1,258 @@
+}
+
+impl<T: Transport> PublicResolver<T> {
+    pub fn new(eth: Eth<T>, resolver_addr: Address) -> Self {
```
```suggestion
    /// Creates new instance of [`PublicResolver`] given contract address.
    pub fn new(eth: Eth<T>, resolver_addr: Address) -> Self {
```

tomusdrw commented on src/contract/ens/reverse_resolver.rs:
```diff
@@ -0,0 +1,47 @@
+}
+
+impl<T: Transport> ReverseResolver<T> {
+    pub fn new(eth: Eth<T>, resolver_addr: Address) -> Self {
```
Missing docs for public entries in this file.

tomusdrw reviewed (commented):
Thank you so much for all the effort and apologies for the delay. Would you mind to fix all the remaining documentation grumbles?

SionoiS commented:
A per [ens/mod.rs](https://github.com/SionoiS/rust-web3/blob/a0f7b5715b2c495609f22c8a2107f07d8a941552/src/contract/ens/mod.rs#L8) the only type accessible is ENS.

@tomusdrw most of the documentation your asking for is not user facing. Are you sure? More docs is better I get it.

tomusdrw commented:
> @tomusdrw most of the documentation your asking for is not user facing. Are you sure? More docs is better I get it.

Correct, internal documentation is useful as well and I think we should put a bit of attention and effort there. Also from my experience it happens rather frequently that things that initially were not public end up being used at some point, so we can already be prepared for this :)
